lib/media/streaming_engine.js

```diff
@@ -2060,17 +2060,33 @@ shaka.media.StreamingEngine = class {
     }
 
     if (this.manifest_.sequenceMode) {
-      const lastDiscontinuitySequence =
-          mediaState.lastSegmentReference ?
-              mediaState.lastSegmentReference.discontinuitySequence : null;
+      const lastDiscontinuitySequence = mediaState.lastSegmentReference ?
+          mediaState.lastSegmentReference.discontinuitySequence : null;
       // Across discontinuity bounds, we should resync timestamps for
       // sequence mode playbacks.  The next segment appended should
       // land at its theoretical timestamp from the segment index.
-      if (reference.discontinuitySequence != lastDiscontinuitySequence ||
-          mediaState.needsResync) {
+      const shouldResync = reference.discontinuitySequence !==
+          lastDiscontinuitySequence || mediaState.needsResync;
+      if (shouldResync) {
         mediaState.needsResync = false;
-        operations.push(this.playerInterface_.mediaSourceEngine.resync(
-            mediaState.type, reference.startTime));
+        operations.push(this.playerInterface_.mediaSourceEngine
+            .resync(mediaState.type, reference.startTime));
+      }
+      // Reset the sequence mode text offset after discontinuity.
+      // Prefer video timestamp for sync when available.
+      const isBestMediaStateForTimestamps = (mediaState.type ===
+        ContentType.VIDEO || !this.mediaStates_.has(ContentType.VIDEO));
+      if (shouldResync && lastDiscontinuitySequence !== null &&
+        isBestMediaStateForTimestamps) {
+        const segment = await this.fetch_(mediaState, reference);
+        const timestamp = this.playerInterface_.mediaSourceEngine
+            .getTimestampAndDispatchMetadata(mediaState.type, segment,
+                reference, mediaState.stream.mimeType, reference.startTime);
+        if (timestamp !== null) {
+          const offset = reference.startTime - timestamp;
+          operations.push(this.playerInterface_.mediaSourceEngine
+              .resetTextSequenceModeOffset(offset));
+        }
```

Comment:
Instead of exposing the `getTimestampAndDispatchMetadata` function, and calculating the offset externally, could we not just reset the `textSequenceModeOffset_` and let it pull the offset from first video segment of the period?